### PR TITLE
Overriding subtask parameters

### PR DESCRIPTION
**Pull request by zemse**
- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [ ] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [x] I didn't do anything of this.

---

This PR adds an optional third argument to the `hre.run` function which enables to override subtask arguments.

### CODE REVIEW

File: packages/hardhat-core/src/types/runtime.ts [hunk 1]
```diff
@@ -183,7 +187,8 @@ export interface TasksMap {
 
 export type RunTaskFunction = (
   name: string,
-  taskArguments?: TaskArguments
+  taskArguments?: TaskArguments,
+  subtaskArguments?: { [subtaskName: string]: TaskArguments }
```
Comment (fvictorio) [inline packages/hardhat-core/src/types/runtime.ts:191]: Shouldn't this be

```suggestion
  subtaskArguments?: SubtaskArguments
```

?

File: packages/hardhat-core/src/internal/core/runtime-environment.ts [hunk 2]
```diff
@@ -334,9 +354,14 @@ export class Environment implements HardhatRuntimeEnvironment {
    */
   private _resolveValidTaskArguments(
     taskDefinition: TaskDefinition,
-    taskArguments: TaskArguments
+    taskArguments: TaskArguments,
+    subtaskArguments: SubtaskArguments = {}
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/core/runtime-environment.ts:358]: This is a private function so the default value is not necessary.

File: packages/hardhat-core/src/types/runtime.ts [hunk 3]
```diff
@@ -146,6 +146,10 @@ export interface TaskDefinition extends ConfigurableTaskDefinition {
  */
 export type TaskArguments = any;
 
+export interface SubtaskArguments {
+  [subtaskName: string]: TaskArguments;
+}
+
 export interface RunSuperFunction<ArgT extends TaskArguments> {
   (taskArguments?: ArgT): Promise<any>;
```
Comment (fvictorio) [inline packages/hardhat-core/src/types/runtime.ts:154]: Should this signature be updated?

File: packages/hardhat-core/src/internal/core/runtime-environment.ts [hunk 4]
```diff
@@ -267,16 +275,28 @@ export class Environment implements HardhatRuntimeEnvironment {
     globalAsAny.runSuper = runSuper;
 
     let modifiedHreWithParentTaskProfile: any | undefined;
-    if (this.hardhatArguments.flamegraph === true) {
+    if (
+      this.hardhatArguments.flamegraph === true ||
+      subtaskArguments !== undefined
+    ) {
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/core/runtime-environment.ts:281]: Hmm, I don't think this is right. This seems to mean that we will generate a flamegraph when subtask arguments are defined, which we don't want to do.

File: packages/hardhat-core/src/internal/core/runtime-environment.ts [hunk 5]
```diff
@@ -218,20 +223,23 @@ export class Environment implements HardhatRuntimeEnvironment {
...
     if (taskDefinition instanceof OverriddenTaskDefinition) {
       runSuperFunction = async (
-        _taskArguments: TaskArguments = taskArguments
+        _taskArguments: TaskArguments = taskArguments,
+        _subtaskArguments: SubtaskArguments = subtaskArguments
       ) => {
         log("Running %s's super", taskDefinition.name);
 
         if (taskProfile === undefined) {
           return this._runTaskDefinition(
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/core/runtime-environment.ts:239]: There is another `this._runTaskDefinition` call after this that is not receiving the `_subtaskArguments`. I can't add a comment on that line, so I'm mentioning it here.

File: packages/hardhat-core/src/internal/core/runtime-environment.ts (hunk 4 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/core/runtime-environment.ts:281]: I guess we can always make the `subtaskArguments !== undefined`, by defaulting to `{}`

File: packages/hardhat-core/src/types/runtime.ts (hunk 3 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/types/runtime.ts:154]: yep

File: packages/hardhat-core/src/types/runtime.ts (hunk 1 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/types/runtime.ts:191]: Good catch. It looks better.

File: packages/hardhat-core/src/internal/core/runtime-environment.ts (hunk 2 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/core/runtime-environment.ts:358]: Agreed. It's not undefined when it's called.

File: packages/hardhat-core/src/internal/core/runtime-environment.ts [hunk 6]
```diff
@@ -260,23 +269,39 @@ export class Environment implements HardhatRuntimeEnvironment {
...
+
     if (this.hardhatArguments.flamegraph === true) {
-      // We create a modified version of `this`, as we want to keep track of the
-      // `taskProfile` and use it as `callerTaskProfile` if the action calls
-      // `run`, and add a few utility methods.
-      //
-      // Note that for this to work we need to set the prototype later
+      // To generate flamegraph, we want to keep track of the `taskProfile` and
+      // use it as `callerTaskProfile` if the action calls `run`, and add a few
+      // utility methods.
```
Comment (alcuadrado) [inline packages/hardhat-core/src/internal/core/runtime-environment.ts:302]:
```suggestion
      // We modify the `this` again to add  a few utility methods.
```

File: packages/hardhat-core/src/internal/core/runtime-environment.ts [hunk 7]
```diff
@@ -260,23 +269,39 @@ export class Environment implements HardhatRuntimeEnvironment {
...
     const previousRunSuper: any = globalAsAny.runSuper;
     globalAsAny.runSuper = runSuper;
 
-    let modifiedHreWithParentTaskProfile: any | undefined;
+    // We create a modified version of `this`, as we want to keep track of the
+    // `subtaskArguments` and use it to generate updated `subtaskArguments` if
+    // the action calls `run`.
+    //
+    // Note that for this to work we need to set the prototype later
+    let modifiedHreWithParentTaskProfile: any = {
```
Comment (alcuadrado) [inline packages/hardhat-core/src/internal/core/runtime-environment.ts:284]:
```suggestion
    let modifiedHre: any = {
```

This name is outdated.

**Review by alcuadrado — APPROVED** [review]

**Pull request merged**